### DISCUSSION TITLE
How do I disable compile-time ability to send text = uuid query?

### DISCUSSION THREAD

**Discussion by k-bx** [SeaQL/sea-query#790] [Q&A] [answered]
One of the most common runtime errors I get is when I send text instead of uuid to my PostgreSQL query, which ends up in a runtime error.

Is there a way to disable this?

Not sure if sea-query or sea-orm question as I mostly use sea-orm.

**Comment by Expurple** [comment, marked as answer]
If I remember everything correctly:
- `sea-query` expressions are untyped by design. And so are `sea-orm` conditions, which are built on them.
- `sea-orm` inserts are done through a typed `ActiveModel` and you shouldn't have such bugs there.
- `sea-orm` updates can be done through a typed `ActiveModel` ([UpdateMany::set](https://docs.rs/sea-orm/latest/sea_orm/query/struct.UpdateMany.html#method.set)) or an untyped expr ([UpdateMany::col_expr](https://docs.rs/sea-orm/latest/sea_orm/query/struct.UpdateMany.html#method.col_expr)). You can use the former where you can.
- Custom `sea-orm` selects (using `into_model` or `into_tuple`, rather than automatically getting a `Model`) are not type checked against your query. It's possible to declare a Rust type (or field name) that doesn't match the query and results in a runtime error.

Untyped exprs is one of the tradeoffs that `sea-query` makes. This makes it easier to compose dynamic queries and probably speeds up compilation. If type errors are a big issue for you and you're in a position where you can still switch the library, you can consider using Diesel. There's a [documentation page](https://www.sea-ql.org/SeaORM/docs/internal-design/diesel/) with the comparison.

Disclaimer: I'm not a maintainer.

**Reply by k-bx** [reply]
Thanks. I've noticed that `mymodel::Entity::find_by_id` is typed, while `.filter(mymodel::Column::Id.eq(val)` is not. Would be nice to add a typed eq, since I'm constructing using SeaORM entities

**Reply by Expurple** [reply]
To have a typed `eq`, each column would need be its own Rust type with SQL type metadata, like in Diesel. But even then, if `sea-query` is kept unchanged, you'd still have a problem that any `sea-query` expression that wraps the column is "type-erased" again and you can do incorrect things like `mymodel::columns::Id.add(1).eq("string")` (where `mymodel::columns::Id.add(1)` evaluates to untyped `SimpleExpr` if the column is an integer, or refuses to compile otherwise)
